### PR TITLE
Extension porting 4.0.2

### PR DESCRIPTION
**Pull request by kumarak**
### Requirements
Migrating the sample osquery extension to build with SDK version 4.0.2. 

### Description of the Change
The PR changes make a hard switch to build with SDK version 4.0.2. The CMake files changes are not compatible with the previous version. Further changes are needed to build it with previous SDK.

### Alternate Designs
NA

### Why Shouldn't This Be In oquery Core?
NA

### Benefits
No new feature added

### Possible Drawbacks
The changes will not build with previous SDK versions. 

### Verification Process
Build and run the existing test suites. 

### Applicable Issues
NA

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,23 +12,36 @@
...
+endfunction()
 
 function(trailofbitsExtensionsMain)
   message(STATUS "================================")
   message(STATUS "Trail of Bits osquery extensions")
-  message(STATUS "Version: $ENV{OSQUERY_EXTENSION_GROUP_VERSION}")
+  message(STATUS "Version: ${OSQUERY_EXTENSION_GROUP_VERSION}")
   message(STATUS "================================")
 
+  generateOsqueryVersion(${OSQUERY_VERSION_AUTODETECTED})
```
Comment (Smjert) [inline CMakeLists.txt]:
```suggestion
  generateOsqueryVersion(${OSQUERY_VERSION_INTERNAL})
```
The autodetected version is only set if the user is not specifying/overriding the version with `-DOSQUERY_VERSION`, `OSQUERY_VERSION_INTERNAL` is what you want instead.

**Comment by JulianCeaser** [comment]
Thank you for creating this branch. I was able to checkout this branch and create extension for fwctl for osquery. But I had to do a couple of changes to get the extension to build.

1. There is no Version.h initially and it seems to be built when I am building osquery but it is building in the osquery-extensions root directory. I had to manually copy it into the fwctl/src/ folder.
2. osquery/build/external/extension_trailofbits/osquery_extension_group_main.cpp:32:10 the function waitForShutdown() needed to be replaced by waitThenShutdown()

Only when I did both of these changes was I able to get the extension to build.

**Comment by GarretReece** [comment]
Building on macos against current osquery master (commit `93ceef7`), this errors when building the `darwin_unified_log` extension, specifically because the `boost/winapi/config.hpp` file isn't included on non-windows platforms (but is still required for `cmd.hpp`). Fixing this in the osquery source tree (modifying the file `libraries/cmake/source/boost/CMakeFiles.txt` to add the include files before existing out of the `generateBoostWinapi` function) exposed an error in boost, specifically in `pipe.hpp:298:14: error: no viable overloaded '=' for _buf = std::move(lhs);`

**Comment by JulianCeaser** [comment]
Is anyone actually able to build this on Windows 10 ?
I created the symlink between the osquery-extensions folder into the osquery externals folder. When I am doing the `cmake -G "Visual Studio 16 2019" -A x64 -T v141 ..` command I can see the libraries are linking
```-- ================================
-- Trail of Bits osquery extensions
-- Version: 1.2
-- ================================
Importing libraries from 'C:/Users/samra/osquery/external/extension_trailofbits/libraries/'
Importing extensions from 'C:/Users/samra/osquery/external/extension_trailofbits'
-- ================================```

But when I am actually doing the build using 
`cmake --build . --config RelWithDebInfo -j10`

I am not seeing any executable being built at all. Can anyone help me understand what I am doing wrong?

**Comment by JulianCeaser** [comment]
Did all the requisite work. But when I am finally trying to build the extensions on my Windows 10 desktop, it is only building the ntfs_forensics and windows_sync_objects extension.

`PS C:\Users\JC\Source\Repos\osquery\build> cmake -G "Visual Studio 16 2019" -A x64 -T v141 ..
-- osquery version: 4.1.1-25-g668a1d86-dirty
-- Build type:
-- Shared libraries: OFF
-- Importing: facebook/markupsafe
-- Importing: facebook/jinja2
-- Importing: source/boost
-- Importing: source/icu
-- Importing: source/bzip2
-- Importing: source/gflags
-- Importing: source/glog
-- Importing: source/googletest
-- Importing: facebook/libarchive
-- Importing: facebook/linenoise-ng
-- Importing: source/rapidjson
-- Importing: source/rocksdb
-- Importing: source/sqlite
-- Disabling the SQLite R*Tree module
-- Disabling the SQLite FTS extensions
-- Importing: source/thrift
-- Importing: source/zlib
-- Importing: source/zstd
-- Importing formula: formula/openssl
--   Version: 1.0.2t
--   Revision: 1
--   Dependencies: thirdparty_zlib
-- Importing: source/aws-sdk-cpp
-- ================================
-- Trail of Bits osquery extensions
-- Version: 1.2
-- ================================
Importing libraries from 'C:/Users/JC/Source/Repos/osquery/external/extension_trailofbits/libraries/'
 > tsk_builtin
Importing extensions from 'C:/Users/JC/Source/Repos/osquery/external/extension_trailofbits'
 > ntfs_forensics
 > windows_sync_objects
-- ================================
-- cppcheck: Enabled with C:/Program Files/Cppcheck/cppcheck.exe
-- Found WIX toolset at: C:/Program Files (x86)/WiX Toolset v3.11
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/JC/Source/Repos/osquery/build`

`PS C:\Users\JC\Source\Repos\osquery\build> $env:TRAILOFBITS_EXTENSIONS_TO_BUILD
windows_sync_objects,fwctl,ntfs_forensics`

Branch information

**Comment by JulianCeaser** [comment]
I have tried on another Windows 10 system and am getting the same issue again. Just cannot get the fwctl extension to get built. Any help on how to fix this ? If I need to give additional information please ask.

**Pull request merged**